server/src/main/java/org/apache/iotdb/db/mpp/aggregation/slidingwindow/SlidingWindowAggregator.java

```diff
@@ -0,0 +1,143 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.mpp.aggregation.slidingwindow;
+
+import org.apache.iotdb.db.mpp.aggregation.Accumulator;
+import org.apache.iotdb.db.mpp.aggregation.Aggregator;
+import org.apache.iotdb.db.mpp.plan.planner.plan.parameter.AggregationStep;
+import org.apache.iotdb.db.mpp.plan.planner.plan.parameter.InputLocation;
+import org.apache.iotdb.tsfile.exception.write.UnSupportedDataTypeException;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.read.common.TimeRange;
+import org.apache.iotdb.tsfile.read.common.block.TsBlock;
+import org.apache.iotdb.tsfile.read.common.block.TsBlockBuilder;
+import org.apache.iotdb.tsfile.read.common.block.column.Column;
+import org.apache.iotdb.tsfile.read.common.block.column.ColumnBuilder;
+import org.apache.iotdb.tsfile.read.common.block.column.TimeColumn;
+
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public abstract class SlidingWindowAggregator extends Aggregator {
+
+  // cached partial aggregation result of pre-aggregate windows
+  protected Deque<PartialAggregationResult> deque;
+
+  public SlidingWindowAggregator(
+      Accumulator accumulator, List<InputLocation[]> inputLocationList, AggregationStep step) {
+    super(accumulator, step, inputLocationList);
+    this.deque = new LinkedList<>();
+  }
+
+  @Override
+  public void processTsBlock(TsBlock tsBlock) {
+    checkArgument(
+        step.isInputPartial(),
+        "Step in SlidingWindowAggregationOperator can only process partial result");
+    TimeColumn timeColumn = tsBlock.getTimeColumn();
+    Column[] valueColumn = new Column[inputLocationList.get(0).length];
+    for (int i = 0; i < inputLocationList.get(0).length; i++) {
+      InputLocation inputLocation = inputLocationList.get(0)[i];
+      checkArgument(
+          inputLocation.getTsBlockIndex() == 0,
+          "SlidingWindowAggregationOperator can only process one tsBlock input.");
+      valueColumn[i] = tsBlock.getColumn(inputLocation.getValueColumnIndex());
+    }
+    processPartialResult(new PartialAggregationResult(timeColumn, valueColumn));
+  }
+
+  @Override
+  public void updateTimeRange(TimeRange curTimeRange) {
+    this.curTimeRange = curTimeRange;
+    evictingExpiredValue();
+  }
+
+  /** evicting expired element in queue and reset expired aggregateResult */
+  protected abstract void evictingExpiredValue();
+
+  /** update queue and aggregateResult */
+  public abstract void processPartialResult(PartialAggregationResult partialResult);
+
+  protected static class PartialAggregationResult {
+
+    private final TimeColumn timeColumn;
+    private final Column[] valueColumns;
```

Comment:
Maybe it should be renamed to `partialResultColumns`. When I read the code, I always thought the valueColumns means the value of origin time series.

Reply:
fixed